File: Src/Controller/NetworkStats.py
from time import time
from typing import NamedTuple

class HealthBasics(NamedTuple):
    count = 0
    last_message_time = time()

class HealthCore(NamedTuple):
    min = float('inf')
    max = -float('inf')
    mean = 0.0
    variance = 0.0
    M2 = 0.0

class NodeReport(NamedTuple):
    packetLoss = 0.0
    latency = HealthCore()
    jitter = HealthCore()
    throughput = HealthCore()

class NetworkStats:
    def __init__(self, _id: int, _members: list) -> None:
        self.id = _id
        self.members = _members
        self.basics = [HealthBasics()] * len(_members)
        self.health_report = [NodeReport()] * len(_members)

    def update(self, id: int, packetSize: int, timestamp: int, sequence_number: int):
        for i in range(len(self.members)):
            if self.members[i] == id:
                node = self.health_report[i]
                basics = self.basics[i]
                now = time()
                basics.count += 1
                self.calculate(node.latency, basics.count, now - timestamp)
                self.calculate(node.jitter, basics.count, node.latency.variance)
                node.packetLoss = ((sequence_number - basics.count) / sequence_number) * 100
                ellapsedSeconds = (now - basics.last_message_time) / 1000
                self.calculate(node.throughput, basics.count, (packetSize * 8) / ellapsedSeconds)
                basics.last_message_time = now

    def calculate(edge: HealthCore, count: int, n: float):
        # From: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
        edge.min = min(edge.min, n)
        edge.max = max(edge.max, n)
        delta = n - edge.mean
        edge.mean += delta / count
        delta2 = n - edge.mean
        edge.M2 += delta * delta2
        edge.variance = edge.M2 / count

File: Src/SSSF/src/NetworkStats/NetworkStats.cpp
#include <Arduino.h>
#include <CANNode/CANNode.h>
#include <PTPClient/PTPClient.h>
#include <NetworkStats/NetworkStats.h>
#include <FlexCAN_T4.h>
#include <TimeClient/TimeClient.h>

NetworkStats::NetworkStats(size_t _size, PTPClient* _timeClient):
    timeClient(_timeClient),
    size(_size),
    Basics(new HealthBasics [_size]),
    HealthReport(new NodeReport [_size])
{}

NetworkStats::~NetworkStats()
{
    delete[] HealthReport;
    delete[] Basics;
}

void NetworkStats::update(uint16_t i, int packetSize, uint64_t timestamp, uint32_t sequenceNumber, int64_t now)
{
    int delay = int64_t(int64_t(now - timestamp) / 1000);
    // if these numbers are 0 this is the first message we've received
    if ((Basics[i].lastMessageTime != 0) && (Basics[i].lastSequenceNumber != 0))
    {
        // Serial.print("Controller Send: ");
        // Serial.print(timestamp);
        // Serial.print(" SSSF Recv: ");
        // Serial.print(_now);
        // Serial.print(" Diff: ");
        // Serial.println(delay);
        calculate(HealthReport[i].latency, abs(delay));
        calculate(HealthReport[i].jitter, HealthReport[i].latency.variance);
        // If no packet loss then sequence number = last sequence number + 1
        int32_t packetsLost = int64_t(sequenceNumber) - (Basics[i].lastSequenceNumber + 1);
        // If packetsLost is negative then this usually indicates duplicate or
        // out of order frame. Etherway not a lost packet.
        HealthReport[i].packetLoss += (packetsLost > 0) ? packetsLost : 0;
        HealthReport[i].goodput += packetSize;
    }
    
    Basics[i].lastMessageTime = now;
    Basics[i].lastSequenceNumber = int64_t(sequenceNumber);
}

void NetworkStats::reset()
{
    delete[] HealthReport;
    HealthReport = new NodeReport[size];
}

void NetworkStats::calculate(struct HealthCore &edge, float n)
{// From: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
    edge.min = min(edge.min, n);
    edge.max = max(edge.max, n);
    edge.count++;
    delta = n - edge.mean;
    edge.mean += delta / edge.count;
    delta2 = n - edge.mean;
    edge.sumOfSquaredDifferences += delta * delta2;
    edge.variance = edge.sumOfSquaredDifferences / edge.count;
}